src/frontend/src/optimizer/optimizer_context.rs

```diff
@@ -107,53 +116,57 @@ impl OptimizerContext {
-            next_plan_node_id: RefCell::new(0),
             sql: Arc::from(""),
             normalized_sql: "".to_owned(),
             explain_options: ExplainOptions::default(),
             optimizer_trace: RefCell::new(vec![]),
             logical_explain: RefCell::new(None),
-            next_correlated_id: RefCell::new(0),
             with_options: Default::default(),
             session_timezone: RefCell::new(SessionTimezone::new("UTC".into())),
-            next_expr_display_id: RefCell::new(0),
             total_rule_applied: RefCell::new(0),
             overwrite_options: OverwriteOptions::default(),
             rcte_cache: RefCell::new(HashMap::new()),
+
+            last_plan_node_id: Cell::new(0),
+            last_correlated_id: Cell::new(0),
+            last_expr_display_id: Cell::new(0),
+
             _phantom: Default::default(),
         }
         .into()
     }
 
     pub fn next_plan_node_id(&self) -> PlanNodeId {
-        *self.next_plan_node_id.borrow_mut() += 1;
-        PlanNodeId(*self.next_plan_node_id.borrow())
-    }
-
-    pub fn get_plan_node_id(&self) -> i32 {
-        *self.next_plan_node_id.borrow()
+        PlanNodeId(self.last_plan_node_id.update(|id| id + 1))
     }
 
-    pub fn set_plan_node_id(&self, next_plan_node_id: i32) {
-        *self.next_plan_node_id.borrow_mut() = next_plan_node_id;
+    pub fn next_correlated_id(&self) -> CorrelatedId {
+        self.last_correlated_id.update(|id| id + 1)
     }
 
     pub fn next_expr_display_id(&self) -> usize {
-        *self.next_expr_display_id.borrow_mut() += 1;
-        *self.next_expr_display_id.borrow()
+        self.last_expr_display_id.update(|id| id + 1)
     }
 
-    pub fn get_expr_display_id(&self) -> usize {
-        *self.next_expr_display_id.borrow()
+    pub(in crate::optimizer) fn backup_elem_ids(&self) -> LastAssignedIds {
+        LastAssignedIds {
+            last_plan_node_id: self.last_plan_node_id.get(),
+            last_correlated_id: self.last_correlated_id.get(),
+            last_expr_display_id: self.last_expr_display_id.get(),
+        }
     }
 
-    pub fn set_expr_display_id(&self, expr_display_id: usize) {
-        *self.next_expr_display_id.borrow_mut() = expr_display_id;
+    /// This should only be called in [`crate::optimizer::plan_node::reorganize_elements_id`].
+    pub(in crate::optimizer) fn reset_elem_ids(&self) {
+        self.last_plan_node_id.set(0);
+        self.last_correlated_id.set(0);
+        self.last_expr_display_id.set(0);
     }
 
-    pub fn next_correlated_id(&self) -> CorrelatedId {
-        *self.next_correlated_id.borrow_mut() += 1;
-        *self.next_correlated_id.borrow()
+    pub(in crate::optimizer) fn restore_elem_ids(&self, backup: LastAssignedIds) {
+        self.last_plan_node_id.set(backup.last_plan_node_id);
+        self.last_correlated_id.set(backup.last_correlated_id);
+        self.last_expr_display_id.set(backup.last_expr_display_id);
```

Comment:
What about just inlining these functions info `reorganize_elements_id`?

Reply:
When adding a new `id` later, we can hardly remember that there's a `reorganize_elements_id` far away need to be updated. So I decided to hide all the ids inside `LastAssignedIds`.

Reply:
I don't get you. We can still add an `id` in `OptimizerContext` but not in `LastAssignedIds`

Reply:
Maybe you should replace the fields to `LastAssignedIds` in `OptimizerContext`

Reply:
The fact is that we missed `last_correlated_id` in `reorganize_elements_id` before. I want to reduce the possibility of future error, so I moved them to `LastAssignedIds` and put it right next to `OptimizerContext`. I also definitely considered directly using `LastAssignedIds` as a field in `OptimizerContext`, but IMO it's too annoying to access these ID fields via a long `self.last_assigned_ids.last_plan_node_id`.

This PR doesn't aim to completely refactor the optimizer element IDs, otherwise I would extract an `IdAllocator` struct to allocate IDs instead of `self.last_plan_node_id.update` in `OptimizerContext`. Also I don't think use interior mutability here in `OptimizerContext` is a good idea. But I don't want to do that in this PR. It's just a quick fix to unblock my future work on other things.